### ISSUE TITLE
v3_40_14_m22 Invalid AxisName lookup

### ISSUE THREAD

**Issue by fjdiazt** [muchimi/JoystickGremlinEx#123]
v3_40_14_m22

I am trying to remap an axis to vJoy Dial 

![image](https://github.com/user-attachments/assets/4481a9b8-92b4-4e40-ad5f-21e524d780f6)

But I am getting:

```
During handling of the above exception, another exception occurred:

 Traceback (most recent call last):
   File "gremlin\ui\ui_common.py", line 726, in _execute_callback
   File "gremlin\ui\ui_common.py", line 599, in get_selection
   File "gremlin\types.py", line 84, in to_enum
 gremlin.error.GremlinError: 'Invalid AxisName lookup, Dial (8)'
```

![image](https://github.com/user-attachments/assets/838df922-38c3-4ab4-8ad9-74ecf92ddc4d)

This works fine in JoystickGremlin

[dill_debug.txt](https://github.com/user-attachments/files/17019742/dill_debug.txt)
[log.txt](https://github.com/user-attachments/files/17019743/log.txt)

**Comment by pavelsubik** [comment]
Hello, is it possible to use all axes? how?

![image](https://github.com/user-attachments/assets/62abdfc8-37d7-4e03-b243-cdc106b8fa81)

Thx

**Comment by muchimi** [comment]
Thank you for reporting. Yes the enum refactor doesn't handle more than the first 8 axes and it looks like you have 16 total selected and vjoy is thus reporting the bottom 8 which have different names.
I will address this in the next patch.

**Comment by muchimi** [comment]
> v3_40_14_m22
> 
> I am trying to remap an axis to vJoy Dial
> 
> ![image](https://private-user-images.githubusercontent.com/14144777/367925469-4481a9b8-92b4-4e40-ad5f-21e524d780f6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY5Nzg2NzAsIm5iZiI6MTcyNjk3ODM3MCwicGF0aCI6Ii8xNDE0NDc3Ny8zNjc5MjU0NjktNDQ4MWE5YjgtOTJiNC00ZTQwLWFkNWYtMjFlNTI0ZDc4MGY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTIyVDA0MTI1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI4N2U3MmRjOWIxYWQzNzBjYWQzY2Q3M2IxZjg5OTNmMzMzYjE4MzRiZWVjMDY4YmViOWU3ODk0MjIyMDExM2YmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.scn5Kdo3VcZsY805CohWyA_7w7NHkwr-F_XYRqSI3Mo)
> 
> But I am getting:
> 
> ```
> During handling of the above exception, another exception occurred:
> 
>  Traceback (most recent call last):
>    File "gremlin\ui\ui_common.py", line 726, in _execute_callback
>    File "gremlin\ui\ui_common.py", line 599, in get_selection
>    File "gremlin\types.py", line 84, in to_enum
>  gremlin.error.GremlinError: 'Invalid AxisName lookup, Dial (8)'
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/14144777/367925192-838df922-38c3-4ab4-8ad9-74ecf92ddc4d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY5Nzg2NzAsIm5iZiI6MTcyNjk3ODM3MCwicGF0aCI6Ii8xNDE0NDc3Ny8zNjc5MjUxOTItODM4ZGY5MjItMzhjMy00YWI0LThhZDktNzRlY2Y5MmRkYzRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTIyVDA0MTI1MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI0MTYxNGU3MDMyOTE3ZjY3MjVjN2ViNGJkNDJhNGJmYjBkMjY1YWExOGM5YjE1MzNjYjU3NTM2YTRjNzY0YTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ummua1zdT_hv6QsH2YDwFgrZUUFWyeAo2pTEfDn0Buw)
> 
> This works fine in JoystickGremlin
> 
> [dill_debug.txt](https://github.com/user-attachments/files/17019742/dill_debug.txt) [log.txt](https://github.com/user-attachments/files/17019743/log.txt)

Thanks for reporting - will address in the next patch.

**Comment by muchimi** [comment]
I did a couple of things as I'm not able to reproduce this with the understanding every system is different.
(1) beefed up the lookup code so it won't blow up if it cannot find a value - it will report the axis name as invalid but won't take the whole system down with it.  This may still create an error, but it will shed some light as to why.
(2) the windows API only supports up to 8 axes, 128 buttons and 4 hats. You should not select more than 8 devices in Vjoy configuration and what's happening here is the bottom 8 are duplicated for the top 8 and just report as different names, but these shouldn't matter as far as I know as only 8 will report and the names in VJOY don't really matter in GremlinEx.

Here's mapping on Dial (axis 8) using the legacy mapper.

![image](https://github.com/user-attachments/assets/0150c048-029d-4be4-9102-a775f81491b0)

**Comment by muchimi** [comment]
Marking resolved as of 13.40.15ex m2 (added boundary check).
Direct Input devices can only have up to 8 axes, 128 buttons and 4 hats in any combination.

**Comment by Festive4020** [comment]
To clarify, can we still USE the remaining options on vjoy so long as we don't select more than 8 axis (so we could have the axis show with the clearer names in software), or do we need to ONLY use the top 8 axis in vjoy? (x-dial/slider2)

**Comment by muchimi** [comment]
GremlinEx (and there is a similar issue reported on the original Gremlin) boils down to this: the Microsoft DirectInput specification only supports up to 8 (eight) axes on a game input device, so a single device cannot have more than 8 axes, 128 buttons and 4 hats.  I don't recall if Vjoy lets you define more than that but you cannot have more than 8 axes at a time on any single VJOY device, and furthermore, each defined VJOY device must be different from the other by either axis count, button count or hat count.
Can you post a screenshot of your VJOY configuration app showing the problematic device?
There may be a bug in that it should not let you use more than 8 even if more than 8 are reported so I'll look into that.  I do know that this error when reported in the past had to do with VJOY configurations and specs but the point is GremlinEx should handle this correctly and point out any configuration problems with the inputs.

**Comment by Festive4020** [comment]
My question is not so much a bug as it is clarification.

With the newer versions of vjoy, we can select up to 8 axis, but you can choose what those axis are called to a degree. Wheel rather than say, X, or Clutch rather than say, rZ, etc...

**Comment by muchimi** [comment]
Currently, GremlinEx expects the default names and the first 8 names as presented.  I tested ok with the registry hack, although it causes some strange behaviors as it confused DirectInput a bit (although it works).
GremlinEx directly uses the information reported by the API and internally uses its own naming for axes (the API actually uses a hardware GUID and an index - no names are used).
I unfortunately don't have a game that only works with a wheel and I was able to test GremlinEx ok with the hack or without so I assume one of the two methods would work here.
